index_configuration_tool/main.py

```diff
@@ -0,0 +1,98 @@
+import sys
+from typing import Optional
+
+import logstash_conf_parser as parser
+import search_endpoint
+import utils
+
+# Constants
+SUPPORTED_ENDPOINTS = ["opensearch", "elasticsearch"]
+HOSTS_KEY = "hosts"
+USER_KEY = "user"
+PWD_KEY = "password"
+
+
+# TODO Only supports basic auth for now
+def get_auth(input_data: dict) -> Optional[tuple]:
+    if USER_KEY in input_data and PWD_KEY in input_data:
+        return input_data[USER_KEY], input_data[PWD_KEY]
+
+
+def get_endpoint_info(plugin_config: dict) -> tuple:
+    endpoint = "https://" if ("ssl" in plugin_config and plugin_config["ssl"]) else "http://"
+    # "hosts" can be a simple string, or an array of hosts for Logstash to hit.
+    # This tool needs one accessible host, so we pick the first entry in the latter case.
+    endpoint += plugin_config[HOSTS_KEY][0] if type(plugin_config[HOSTS_KEY]) is list else plugin_config[HOSTS_KEY]
+    endpoint += "/"
+    return endpoint, get_auth(plugin_config)
+
+
+def fetch_all_indices_by_plugin(plugin_config: dict) -> dict:
+    endpoint, auth_tuple = get_endpoint_info(plugin_config)
+    return search_endpoint.fetch_all_indices(endpoint, auth_tuple)
+
+
+def validate_plugin_config(config: dict, key: str):
+    # The value of each key is a list of plugin configs.
+    # Each config is a tuple, where the first value is the endpoint type.
+    supported_endpoint = next((p for p in config[key] if p[0] in SUPPORTED_ENDPOINTS), None)
+    if not supported_endpoint:
+        raise ValueError("Could not find any supported endpoints in section: " + key)
+    plugin_config = supported_endpoint[1]
+    if HOSTS_KEY not in plugin_config:
+        raise ValueError("No hosts defined for endpoint: " + supported_endpoint[0])
+    if USER_KEY in plugin_config and PWD_KEY not in plugin_config:
+        raise ValueError("Invalid auth configuration (no password for user) for endpoint: " + supported_endpoint[0])
+    elif PWD_KEY in plugin_config and USER_KEY not in plugin_config:
+        raise ValueError("Invalid auth configuration (Password without user) for endpoint: " + supported_endpoint[0])
+
+
+def validate_logstash_config(config: dict):
+    if "input" not in config or "output" not in config:
+        raise ValueError("Missing input or output data from Logstash config")
+    validate_plugin_config(config, "input")
+    validate_plugin_config(config, "output")
+
+
+def print_report(source: dict, target: dict) -> set:
```

Comment:
Minor naming note: this does print a report, but the more important job seems to be determining which indices to create--true?
Maybe either break it into two functions with clear jobs or come up with a name that encompasses both jobs?

Reply:
👍  Separated into `get_index_differences` and `print_report` now